compiler/sem/semtypes.nim

```diff
@@ -1717,21 +1732,27 @@ template modifierTypeKindOfNode(n: PNode): TTypeKind =
   else: tyNone
 
 proc semTypeClass(c: PContext, n: PNode, prev: PType): PType =
+  addInNimDebugUtils(c.config, "semTypeClass", n, prev, result)
   # if n.len == 0: return newConstraint(c, tyTypeClass)
   let
     pragmas = n[1]
     inherited = n[2]
 
   result = newOrPrevType(tyUserTypeClass, prev, c)
   result.flags.incl tfCheckedForDestructor
-  var owner = getCurrOwner(c)
-  var candidateTypeSlot = newTypeWithSons(owner, tyAlias, @[c.errorType], c.idgen)
+  let owner = getCurrOwner(c)
+  var
+    candidateTypeSlot = newTypeWithSons(owner, tyAlias, @[c.errorType], c.idgen)
+    cycleDetector = initIntSet()
   result.sons = @[candidateTypeSlot]
   result.n = n
 
   if inherited.kind != nkEmpty:
     for n in inherited.sons:
-      result.add semTypeNode(c, n, nil)
+      let t = semTypeNode(c, n, nil)
+      if isRecursiveType(t, cycleDetector):
+        c.config.localReport(n.info, reportTyp(rsemIllegalRecursion, t))
```

Comment:
It's not strictly necessary to check for `t == prev` here, as `t` is not added via `rawAddSon`.

Reply:
Couldn't one do `Foo = object of Foo` in which case we'd have the same `tyForward`?

Reply:
Yep, that case is possible, but the recursion isn't directly fatal and already rejected in `typeSectionFinalPass`.

Nonetheless, an `inheritance only work with non-final object` error is currently produced with your example, as `semObjectNode` doesn't properly consider `tyForward`.